### ISSUE TITLE
NPM audit dependency high severity findings

### ISSUE THREAD

**Issue by murdocha** [julianhille/MuhammaraJS#108]
Running:
```npm audit```
returns two high severity findings.

Not sure if you have plans to update these dependencies on the node-pre-gyp -> tar module?

NPM audit findings with muhammara version 1.10.0 from npm:

```
High            Arbitrary File Creation/Overwrite due to insufficient         
                  absolute path sanitization                                    

  Package         tar                                                           

  Patched in      >=3.2.2 <4.0.0 || >=4.4.14  <5.0.0 || >=5.0.6 <6.0.0 ||       
                  >=6.1.1                                                       

  Dependency of   muhammara                                                     



  Path            muhammara > node-pre-gyp > tar

  More info       https://npmjs.com/advisories/1770


  High            Arbitrary File Creation/Overwrite via insufficient symlink
                  protection due to directory cache poisoning

  Package         tar

  Patched in      >=3.2.3 <4.0.0 || >=4.4.15  <5.0.0 || >=5.0.7 <6.0.0 ||
                  >=6.1.2

  Dependency of   muhammara

  Path            muhammara > node-pre-gyp > tar

  More info       https://npmjs.com/advisories/1771
  ```

https://www.npmjs.com/advisories/1771
"The tar package has a high severity vulnerability before versions 3.2.3, 4.4.15, 5.0.7, and 6.1.2."

https://www.npmjs.com/advisories/1770
"The tar package has a high severity vulnerability before versions 3.2.2, 4.4.14, 5.0.6, and 6.1.1."

**Comment by murdocha** [comment]
ok. not sure what changed, but I'm not seeing the issue any more. Maybe I just needed to delete node_modules and run install again...

**Comment by murdocha** [comment]
Okay, I think know _why_ the error went away (and why it might still be an issue for others).

_IF_ you install from the published npm package (v 1.10.0, https://www.npmjs.com/package/muhammara), you will see the two high severity audit findings listed above. 
However, _if_ you install from the GitHub repo link directly (something I'd been testing when the audit findings went away), with a package.json reference like this:
```
    "muhammara": "git://github.com/julianhille/MuhammaraJS.git",
```
Then the two high severity npm audit findings disappear (on my machine at least. I'm running local Node version 14LTS (v.14.17.4) on Windows 10. NPM version (v.6.14.14).

I'm not confident about _why_ this difference exists, but I'm guessing that this has to be a dependency issue, and is likely due to the "bundledDependencies".

```
  "dependencies": {
    "node-pre-gyp": "^0.17.0"
  },
  "bundledDependencies": [
    "node-pre-gyp"
  ],
  ```
I'm thinking that the "direct" GitHub link doesn't use the older "bundled" dependency that is probably included with the NPM module install package.

With the "npm install" method I see these errors in my local terminal (where "pii-scan-api" is the name of my local app):

```
$ npm install

> dtrace-provider@0.8.8 install C:\projects\pii-scan-api\node_modules\dtrace-provider
> node-gyp rebuild || node suppress-error.js

C:\projects\pii-scan-api\node_modules\dtrace-provider>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild ) 
gyp ERR! UNCAUGHT EXCEPTION 
gyp ERR! stack Error: spawn EPERM
gyp ERR! stack     at ChildProcess.spawn (internal/child_process.js:403:11)
gyp ERR! stack     at spawn (child_process.js:667:9)
gyp ERR! stack     at execFile (child_process.js:319:17)
gyp ERR! stack     at regGetValue (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\util.js:27:17)
gyp ERR! stack     at search (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\util.js:51:5)
gyp ERR! stack     at VisualStudioFinder.regSearchKeys (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\util.js:57:3)
gyp ERR! stack     at VisualStudioFinder.findOldVS (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:369:10)
gyp ERR! stack     at VisualStudioFinder.findVisualStudio2015 (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:337:17)
gyp ERR! stack     at C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:66:12
gyp ERR! stack     at VisualStudioFinder.parseData (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:236:5)
gyp ERR! System Windows_NT 10.0.18363
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\projects\pii-scan-api\node_modules\dtrace-provider
gyp ERR! node -v v14.17.4
gyp ERR! node-gyp -v v5.1.0
gyp ERR! This is a bug in `node-gyp`.
gyp ERR! Try to update node-gyp and file an Issue if it does not help:
gyp ERR!     <https://github.com/nodejs/node-gyp/issues>
```  
  
I'm not familiar with the gyp module, though I gather it is used to compile C or C++ source code? 

So I _do_ have the GitHub direct link workaround, but ideally, we'd be able to reference the npm module instead.
Thanks!

**Comment by julianhille** [comment]
Thanks for reporting.
Interesting that i did not get an email.
Anyway: 
- node tar could be upgraded but is not a direct dependency so i need to have a update node - pre gyp. (one of the node-tar versions 3.2.3 could be used as this does not raise our node version above the prebuilts im actually building here, so that would be fine to do.)
- with the next node-pre-gyp version we would leave some older node versions behind.
 i was afraid to do so. 
 
 BUT this is an security issue which could be ignored when installing from npm but i see your point. I personally could overwrite the github releases (or someone with access) and use that bug to compromise your machine.
 If you are building yourself then this issue could be safely ignored.
 
 So the only way is to leave ALL musl / node / electron versions with a node version <= 8 behind.
 
 There is already a ticket which would solve this one.
 I will have a look at download stats of the last releases and node versions and will come back to you.
 Is this fine for you?
 
 Cheers.

**Comment by murdocha** [comment]
Thanks for the prompt response!
I'll look for an update to see what our path forward should be (fork, perhaps temporarily, to our own repo, or hold tight for an updated npm module).

**Comment by murdocha** [comment]
Any news on this? 
In your npm download stats investigation, how many users might be affected by a potential change to use the newer mapbox-maintained node-pre-gyp library?
I think this is the earlier issue that referenced plans for that change:
https://github.com/julianhille/MuhammaraJS/issues/71

Thanks!

**Comment by julianhille** [comment]
The issue is a non issue as stated before. If one would want to compromise your system through muhammara and has access to the releases (where the tar file reside) it would be also possible to replace the binary and make it execute some other code.

Anyway plan is to release next version with some new electron versions and then add a 2.0.0 where i remove all node and electron versions not being compatible to mapbox pre gyp

**Comment by julianhille** [comment]
Solved with 2.0 and #128
